### PR TITLE
Upload some scripts about VOC dataset

### PR DESCRIPTION
**Pull request by TaoXieSZ**
1. two for Downloading, 
2. one for splitting train/test set, 
3. one for organizing,
4. yaml files

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Added script and configuration for downloading and setting up the PASCAL VOC dataset.

### 📊 Key Changes
- A new shell script `get_voc.sh` is added to automate the downloading, extraction, and preparation of PASCAL VOC 2007 & 2012 datasets.
- Script includes functionality to handle optional download directory, download data, extract it, and split it into proper dataset format required for YOLOv5.
- A new dataset configuration file `voc.yaml` is added specifying paths, class count, and class names for the VOC dataset in YOLOv5 format.

### 🎯 Purpose & Impact
- 💡 The purpose is to streamline the process of setting up the VOC dataset for users wanting to train YOLOv5 on this dataset.
- 🚀 Impact is a significant reduction in manual setup time and potential for errors, making it easier and more accessible for both new and existing users to work with the VOC dataset.

### CODE REVIEW

**Comment by pppppM** [comment]
Thank you for your contribution！ Could you show the AP on VOC?

**Comment by TaoXieSZ** [comment]
> Thank you for your contribution！ Could you show the AP on VOC?

I think it can be easily obtain over 0.85. I just play with it once.

**Comment by pppppM** [comment]
if I use the default hyper parameters to train,I can only get 0.78（yolov5l） and  0.73(yolov5s).Do you have some suggestions?

**Comment by TaoXieSZ** [comment]
> if I use the default hyper parameters to train,I can only get 0.78（yolov5l） and 0.73(yolov5s).Do you have some suggestions?

You can try focal loss or use bigger models.

**Comment by TaoXieSZ** [comment]
@glenn-jocher Sorry for my commit. I just want to make it easy for kaggle usage.

**Comment by glenn-jocher** [comment]
@ChristopherSTAN ah, figured it might have been a mistake. I still want to get this in, just not sure how to do it in the cleanest way.

**Comment by glenn-jocher** [comment]
@ChristopherSTAN ideally we'd have just 2 files, one to download and format, and the other voc.yaml. This is how we do coco for example. Maybe I could format the labels offline and host them on our google drive for download, and combine that with the voc image download hosted elsewhere. This is what we do for downloading coco for example if you look at the download_coco.sh.

EDIT: I mean get_coco2017.sh

**Comment by TaoXieSZ** [comment]
> @ChristopherSTAN ah, figured it might have been a mistake. I still want to get this in, just not sure how to do it in the cleanest way.

Understand. Maybe I can try to shorten the scripts, like combing downloading, splitting and organizing into one file.
Later for it LOL. I am struggling for interview now.

**Comment by TaoXieSZ** [comment]
@glenn-jocher Working on it. I can complete today.

**Comment by TaoXieSZ** [comment]
@glenn-jocher Check the latest files please~. It is so happy to contribute!

**Comment by glenn-jocher** [comment]
Wow! Awesome, this looks good.

**Comment by glenn-jocher** [comment]
Ok I'm testing this out. Very good bash file coding skills you have there by the way! I opened up a Colab notebook and used these commands. Maybe we can create a VOC tutorial to add to the wiki. Here are my steps:

## Initial Setup
```bash
!git clone https://github.com/ultralytics/yolov5  # clone repo
!pip install -r yolov5/requirements.txt  # install dependencies
%cd yolov5

import torch
from IPython.display import Image, clear_output  # to display images
from utils.google_utils import gdrive_download  # to download models/datasets

clear_output()
print('Setup complete. Using torch %s %s' % (torch.__version__, torch.cuda.get_device_properties(0) if torch.cuda.is_available() else 'CPU'))
```

## 1. Download VOC
```bash
%cd ..
!bash yolov5/data/get_voc.sh
```
Ok so this code runs without error, but then I don't see any new folder in Colab:
Screen Shot 2020-07-04 at 4 33 12 PM

Do you think this command is removing the dataset at the end of get_voc.sh?
`rm -rf ../data
`
## 2. Train a few epochs.



## 3. Visualize

**Comment by TaoXieSZ** [comment]
@glenn-jocher It is used to remove the raw dataset, if it is needed.

**Comment by glenn-jocher** [comment]
Ah, ok, I see. It looks like I should run the script from inside yolov5/. Ok I'll run again now:

## Initial Setup
```bash
!git clone https://github.com/ultralytics/yolov5  # clone repo
!pip install -r yolov5/requirements.txt  # install dependencies
%cd yolov5

import torch
from IPython.display import Image, clear_output  # to display images
from utils.google_utils import gdrive_download  # to download models/datasets

clear_output()
print('Setup complete. Using torch %s %s' % (torch.__version__, torch.cuda.get_device_properties(0) if torch.cuda.is_available() else 'CPU'))
```

## 1. Download VOC
```bash
!bash ./data/get_voc.sh
```
Screen Shot 2020-07-04 at 4 52 46 PM

## 3. Update 20-class model

## 3. Train a few epochs.
```bash
!python train.py --data voc.yaml --cfg yolov5s_20cls.yaml --weights yolov5s.pt --batch 32
```

## 4. Visualize
`train_batch0.jpg
`train_batch0 (1)

**Comment by glenn-jocher** [comment]
It works!! I'm training a few epochs now from pretrained yolov5s.pt. I'll update in a bit.

**Comment by TaoXieSZ** [comment]
@glenn-jocher Glad to see that! So I told you, I am a BIG fan of your team!

**Comment by glenn-jocher** [comment]
One thing I think we should change is the ../data folder that is used for downloads. It's a common name so if someone has an existing ../data folder this download script will delete it (!). 

What do you think if we rename it to `../tmp_voc_download` for example?

**Comment by glenn-jocher** [comment]
Training and testing works correctly. mAP after 1 epoch is 0.306. Great!

Screen Shot 2020-07-04 at 5 07 47 PM

**Comment by TaoXieSZ** [comment]
@glenn-jocher Emm, you point is totally right. That's my fault. I consider only one dataset. I think we can just remove the VOCdevit inside the ../data.

**Comment by TaoXieSZ** [comment]
replace the origin
`
rm -rf ../data
`
with:
`
rm -rf ../data/VOCdevkit
`
`
rm -rf ../data/*.txt
`
I think it matchs your thought!

**Comment by glenn-jocher** [comment]
@ChristopherSTAN ok, all set. I've renamed /data to /tmp, and added a confirmation at the end that download is complete.

I've also updated bb3c34691639b2aafb1a1f567e1dfb551a9a2563 the train script so that the model inherits the dataset class count in the case of any differences between model.yaml and dataset.yaml, so now you don't actually need to update any model.yaml, you can just train directly with yolov5s.yaml:
```bash
!python train.py --data voc.yaml --cfg yolov5s.yaml --weights yolov5s.pt --batch 32
```

**Comment by glenn-jocher** [comment]
@ChristopherSTAN was training on VOC a bit. Got a quick question. Our dataset looks like this:
https://github.com/ultralytics/yolov5/blob/f796708b500d7fa885eaf11cd8d47e8094c1cd40/data/voc.yaml#L11-L13

I see papers like [this](https://www.researchgate.net/figure/Detection-results-on-PASCAL-VOC-dataset-All-models-are-pre-trained-on-MS-COCO-and_tbl2_321180719) though, that show two different test sets, 2007 and 2012. What test set do we have?

Second question is what is a good mAP? I trained YOLOv5s from scratch to 77, from pretrained to 82, and YOLOv5m from pretrained to about 86 mAP@0.5. These all train very fast on Colab Pro, <24h so they can be trained in one go without resuming.
Screen Shot 2020-07-14 at 11 04 41 PM

**Comment by glenn-jocher** [comment]
Here's a closeup.
Screen Shot 2020-07-14 at 11 08 22 PM

**Comment by TaoXieSZ** [comment]
@glenn-jocher I got a leaderboard here.
http://host.robots.ox.ac.uk:8080/leaderboard/main_bootstrap.php

**Comment by TaoXieSZ** [comment]
@glenn-jocher I find another good summary: https://github.com/hoya012/deep_learning_object_detection

Sorry for my inability of offering other help. I am just a noob in front of you...

**Comment by TaoXieSZ** [comment]
@glenn-jocher Besides, I have an idea to implement DIOU-NMS. Do you have a try on it?

**Comment by glenn-jocher** [comment]
@ChristopherSTAN for a full comparison of NMS techniques this repo has my highest recommendation: https://github.com/Zzh-tju/ultralytics-YOLOv3-Cluster-NMS

These are the original authors of CIoU/DIoU. They used ultralytics/yolov3 and applied various NMS techniques and displayed the results in a table with their conclusions added. I've reached out to them for any recommendations they may have for us https://github.com/Zzh-tju/ultralytics-YOLOv3-Cluster-NMS/issues/2

**Comment by TaoXieSZ** [comment]
@glenn-jocher I have given it a try, but find it is very easy to run out of all the CUDA memory.... It seems that we should set a high threshold

**Comment by glenn-jocher** [comment]
@ChristopherSTAN yes, I think this is because you are using focal loss and producing very high recall (and very low precision). 

If your dataset has many classes, one of the main FP multipliers in this setting will be multi-label nms:https://github.com/ultralytics/yolov5/blob/37acbdc0b6ef8c3343560834b914c83bbb0abbd1/utils/utils.py#L566

This allows a single box to be assigned to multiple classes by duplicating it. So for example if a box scores above threshold for both car and truck class, then the box is copied into two boxes, one car, one truck. This helps mAP a bit, but in early training, especially with focal loss it can cause many problems by producing tens of thousands of FPs. I'm open to ideas to make the NMS more robust to this problem, just not sure what the best approach is.

Note this is mainly a focal loss problem, most models and datasets work completely fine as is.

**Comment by glenn-jocher** [comment]
To help understand the scope of the problem, a typical YOLOv5 model might be capable of generating say 25,000 boxes. With COCO and multi-label on, the same model is now capable of generating 25k*80 = **2M boxes (!!), which all must pass through NMS**. So hyp settings that strongly push the model towards recall and away from precision can quickly saturate the system in early training.

**Comment by TaoXieSZ** [comment]
@glenn-jocher I try cluster-weighted-diounms on a 14-class TV show dataset. It seems diounms do not help improve.
Besides, here is my little experiment on VOC dataset with yolov4.:

image
It seems the CIOU loss + diounms play the best performance. 
So, it is still magic for me...

**Comment by glenn-jocher** [comment]
Hmm, maybe it's time to move the repo to CIoU then.

**Comment by TaoXieSZ** [comment]
@glenn-jocher In papers, CIoU is the best...

Let me try on another dataset. I can tell you my result tomorrow. LOL

**Comment by TaoXieSZ** [comment]
@glenn-jocher Start a new game on a new dataset:
If we need to do with small objects and overlapping objects. Can you give me some advices?
image

**Comment by glenn-jocher** [comment]
@ChristopherSTAN you can modify the `--iou-thresh` argument to allow for greater overlap in detections.
`python test.py --iou 0.9`

This sort of dataset may benefit from full rotation, hyp degrees=180:
https://github.com/ultralytics/yolov5/blob/90e7bfdf7561b52433ae2974445c4eaec22ad536/train.py#L39

Or if not rotation then definitely vertical flips:
https://github.com/ultralytics/yolov5/blob/90e7bfdf7561b52433ae2974445c4eaec22ad536/utils/datasets.py#L540-L546

This made me remember I need to integrate the flip probabilities into the hyp dict.

**Comment by TaoXieSZ** [comment]
@glenn-jocher Gotcha, I am so appreciated!

**Comment by glenn-jocher** [comment]
@ChristopherSTAN ok, I finished VOC training tests. YOLOv5l trains to 0.87 and YOLOv5x trains to 0.88 mAP, both in about 80 epochs starting from the official pretrained weights. Starting from scratch produces worse performance, maybe 0.05-0.10 less. 

I was thinking since VOC benefits so much from pretrained weights maybe I should turn around and train COCO starting from the pretrained weights, though this is a bit strange since the pretrained weights themselves trained from scratch... 

In ultralytics/yolov3 I actually observed worse performance when starting from the pretrained darknet backbones than from scratch (and now all of the official weights are trained from scratch with no pretraining or tricks), but I actually never thought to train an ultralytics model starting from a pretrained ultralytics model. Maybe this would help on COCO, and if it does not, I wonder what conclusion could be drawn from that? It's a worthy experiment, I'll run it this week I think.

**Comment by TaoXieSZ** [comment]
@glenn-jocher Can't wait to see your latest finding.

I find a novel data augmentation I've never used: https://github.com/google-research/augmix

I am planning to plug it into my yolo experiment. If it worked, I will tell you immediately!

**Comment by TaoXieSZ** [comment]
@glenn-jocher 
Hi bro.
I just give my stupid idea: train a larger model than yolov5x on a small dataset, because of nan loss

I nearly encounter bottleneck with as many as bag of freebies I can add.

I wonder if there is tricks of tightly modify models to detect smaller objects or overlapping objects.

And I still want to benefit from pretrained model of yours.

Here's some feature network to fuse multi-scale features.
image
I find there is PAN in this repo, and I want to ask, if I change to the others or add other modules, how can I still make use of your pretrained model.

Thanks a lot.

**Comment by glenn-jocher** [comment]
@ChristopherSTAN pretrained weights will be loaded for any layers with **matching names and shapes**, so if you modify yolov5x.yaml's head, and then try to train that with `--weights yolov5x.pt`, most of the pretrained weights in the backbone should still transfer.

**Comment by TaoXieSZ** [comment]
@glenn-jocher Cool.

I really need to read Your way to implement loading_weight() carefully.

Thanks!

**Comment by abhiagwl4262** [comment]
@glenn-jocher @ChristopherSTAN  Hey, Have you ever trained the backbones on ImageNet dataset? Because I think, backbone trained on imagenet like noisy and big datasets might work better as pretrained model.

**Comment by glenn-jocher** [comment]
@abhiagwl4262 no, I have not. Do you mean training on imagenet for classification?

Classification is something I've been wanting to integrate into this repo, but it's a bit of work. We would need a dedicated dataloader and different loss function, in addition to the model modifications. I've taken token steps in this direction, for example with the addition of a classifier head module. If you'd like to help out in the effort, we need all the volunteer help we can get 👍 

https://github.com/ultralytics/yolov5/blob/5a9c5c1d3b02c8796e6af153d57af5821d8db0d6/models/common.py#L108-L118

**Comment by abhiagwl4262** [comment]
@glenn-jocher I will start working in this direction and will let you know once things start to work. What kind of training resources you have anyway ?

There is one more question If you can shed some lights on it. Have you done experiments by keeping number of groups >1 ? 
 And what's your thought on it?
